src/op_analytics/cli/subcommands/pulls/agora/delegate_events.py

```diff
@@ -0,0 +1,143 @@
+from dataclasses import dataclass
+from op_analytics.cli.subcommands.pulls.agora.data_access import (
+    Agora,
+    write,
+    _camelcase_to_snakecase,
+    parse_isoformat_with_z,
+)
+from op_analytics.coreutils.logger import structlog
+from op_analytics.coreutils.request import get_data
+from op_analytics.coreutils.threads import run_concurrently_store_failures
+
+from typing import Any
+import requests
+import pandas as pd
+import os
+
+log = structlog.get_logger()
+
+BASE_URL = "https://vote.optimism.io/api/v1"
+DELEGATES_ENDPOINT = f"{BASE_URL}/delegates"
+API_KEY = os.environ["AGORA_API_KEY"]
+
+
+@dataclass
+class AgoraDelegateVotes:
+    votes_df: pd.DataFrame
+
+
+@dataclass
+class PaginatedResponse:
+    has_next: bool
+    next_offset: int
+    data: Any
+
+
+@dataclass
+class SimplePaginator:
+    url: str
+    limit: int = 50
+
+    def request(self, offset):
+        session = requests.Session()
+        result = get_data(
+            session,
+            url=self.url,
+            headers={"Authorization": f"Bearer {API_KEY}"},
+            params={"offset": offset, "limit": self.limit},
+        )
+
+        assert "meta" in result
+        assert "data" in result
+
+        return PaginatedResponse(
+            has_next=result["meta"]["has_next"],
+            next_offset=result["meta"]["next_offset"],
+            data=result["data"],
+        )
+
+    def fetch_all(self):
+        offset = 0
+        all_data = []
+
+        while True:
+            response = self.request(offset)
+            all_data.extend(response.data)
+
+            if not response.has_next:
+                break
+
+            offset = response.next_offset
+
+        return all_data
+
+
+def fetch_event_data(delegates: list, endpoint: str, workers: int) -> pd.DataFrame:
+    def fetch_delegate_data(address: str, endpoint: str) -> list:
+        paginator = SimplePaginator(url=f"{BASE_URL}/delegates/{address}/{endpoint}", limit=50)
+        try:
+            return paginator.fetch_all()
+        except requests.exceptions.HTTPError as e:
```

Comment:
If we catch the exception then the error will not be detected as a failure by `run_concurrently_store_failures`, I think we have to let exceptions bubble up.